man/vacs.8
.\"                              hey, Emacs:   -*- nroff -*-
.\" quakeforge is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH VACS 8 "2003-07-14"
.\" Please update the above date whenever this man page is modified.
.SH NAME
vacs \- VREng Address Cache Server
.SH SYNOPSIS
.B vacs
.RI [ options ]
.SH DESCRIPTION
\fBVacs\fP (VREng Address Cache Server) keeps dynamicaly the bindings between
world names and there channel strings under the form "multicast_addr/port/ttl".
The main function of vacs is to answer by a \fBchannel\fP string to the
requestor when it receives a request of type "\fBresolve\fP my_world_url".
.PP
The cache contains Ascii entries easily editable by hand.
An entry is composed by an \fBurl\fP (world destription url),
a \fBworldname\fP (extracted from the url),
an \fBuniverse\fP which is the http host (extracted from the url),
and the \fBchannel\fP string computed to be unique "multicast_addr/port/ttl".
.PP
After initialization by a list of well-known couples of world-channel,
vacs maintains in its cache, unique IP multicast addresses and
unique associated ports for a given world.
.PP
The other operations which can do the requestor are "\fBlist\fP" to get all
the couples (world_url - channel) kept in the cache, and "\fBrelease\fP myworld"
to remove an entry from the cache, only granted if the associated universe
matches the requestor's universe.
.SH OPTIONS
The \fBvacs\fP command line accepts the following options:
.TP
.BR  " -d vacs_directory"
Specify where is located the directory which contains vacs cache.
.TP
.BR  " -D"
Debugging mode.
.TP
.BR  " -v"
Prints current version and exits.
.TP
.BR  " -h"
Help message.
.TP
.BR  " -l logfile"
Indicates which log file to use, if not specified default is /var/log/vacs.log.
.SH "SEE ALSO"
.BR vreng (1)
.SH LICENSE
.PP
\fBvacs\fP is an Open Source Software under the GPL (Gnu Public License).
.SH AUTHORS
.PP
Philippe Dax, <dax@enst.fr>.


ChangeLog
Last changes in VREng
---------------------

------------------------------------------
The rules for this file:
* entries are sorted newest-first.
*
* ++ : major
* == : medium
* -- : minor
* Version: major.medium.minor
* date and author format: (YYYY/MM/DD, Author).
------------------------------------------

++ 8
== 8.3
-- 8.3.1
	- vreng homed at github
	- changed CXXFLAGS to avoid warnings under gcc-6.1.1
	- added view.xpm into gui/xpm/Makefile.am
	- moved acinclude.m4 into support/
	(2016/08/07, Philippe Dax)

-- 8.3.0
	- added checklib libsndfile into configure.ac
	- added openal support --with-openal into configure.ac
	(2016/03/26, Philippe Dax)

== 8.2
-- 8.2.2
	- renamed missile to projectile
	- added view third person button in panels.cpp
	- added go button in xidgets.cpp dialog.cpp
	- cleans up bone files
	- fixed endianess in hairs.cpp
	- fixed regression on flashy objects
	- rewriting Render::solids
	- fixed sun rendering
	- added bindSocket in socket.cpp
	- added trace2 without cr in log.cpp
	- added stringcmp in str.cpp
	(2016/01/20, Philippe Dax)

-- 8.2.1
	- fixed regression of ubit/uconf.cpp
	(2015/12/29, Philippe Dax)

-- 8.2.0
	- updated ubit to be in conformity with the clang apple compiler
	- updated check libmysqlclient for Linux
	(2015/10/26, Philippe Dax)

== 8.1
-- 8.1.0
	- added tar.xz distribution
	- added vmm branch
	(2013/10/15, Philippe Dax)

== 8.0
-- 8.0.0
	- added bsphere bounding sphere
	- fixed draw bbox
	- fixed view source using cache
	- added object firework
	- added src/vreng/vector3.hpp
	- deleted src/wo/grid.cpp
	- deleted src/wo/netw.cpp
	- moved src/wo/list.{cpp,hpp} to olist.{cpp,hpp}
	- updated vreng data version to v8
	(2012/09/03, Philippe Dax)

=======
++ 7
== 7.4
-- 7.4.0
	- added src/sys
	- rewriting src/app/* with classes
	- updated display.cpp displayMirrored without pushing names in the buffer
	- fixed xpm.cpp
	- updated src/ubit/uconf.cpp for Type1 fonts under Linux
	- moved src/ubit/nat/*pp to src/ubit
	- moved src/vrel src/vred src/tool to src/util
	- moved src/bone/*pp to src/model
	- updated conf/vreng.spec to build rpm package
	- fixed bin/Makefile.am for vred and vrel
	(2012/07/08, Philippe Dax)

== 7.3
-- 7.3.3
	- moved src/net/md5* src/net/random.cpp to src/vreng/
	- resolved opens/closes leak
	- updated textureList to STL
	- moved src/net/nsl* src/net/socket* to src/vreng/
	- updated file.{ch}pp : added openFile() and closeFile()
	- updated fileio.{ch}pp to file.{ch}pp
	- updated socket.cpp : closeDatagram(), closeStream()
	- updated prof.cpp : added statIO()
	(2012/06/06, Philippe Dax)

-- 7.3.2
	- updated mobileList, stillList, txfList to STL
	- updated txf to be cached
	- added net/noid.{ch}pp comeback
	- updated netobjectList, channelList to STL
	- updated cartList, wearList, lightList, deleteList, invisibleList to STL
	- updated oclass from wclass
	- updated lwo to class
	- updated smoke
	(2012/06/03, Philippe Dax)

-- 7.3.1
	- all remote data are now cached (off, lwo, dxf, cset, book, bodyparts, walls)
	- added vreng/cache.{ch}pp
	- added vreng/reader.{ch}pp
	- added option -r,--refresh
	- added tools obj2parts, jpobj.c, jpoff.c
	- added model/bap.t
	- updated android, body, bap
	- added src/tool
	- added sound.{ch}pp
	(2012/04/25, Philippe Dax)

-- 7.3.0
	- updated vreng-data to version 7.3
	- updated paralellipedic sizes
	(2012/02/24, Philippe Dax)

== 7.2
-- 7.2.8
	- added object smoke
	- moved anim.{ch}pp to animator.{ch}pp
	- added vred branch
	- added vrel branch
	(2012/02/21, Philippe Dax)

-- 7.2.7
	- moved Attractor to Slider
	- moved Repulsor to Bumper
	- moved Follower to Accessory
	(2012/01/29, Philippe Dax)

-- 7.2.6
	- updated object ball from file
	- added object follower
	- added draw bbox
	- added object repulsor
	- added object attractor
	- added dtd menu
	- updated rendering order
	- added model/bvh.{ch}pp
	(2012/01/23, Philippe Dax)

-- 7.2.5
	- added mode roll and tilt to mirage
	- added bubble solid translucid obloid
	- updated plane multi bounding boxes
	- added legend attribute to web
	- added pivot action to web
	- added AVI video reader/writer
	- added PSD image loader
	(2011/12/24, Philippe Dax)

-- 7.2.4
	- moved mirror algos to display.cpp
	- added model helicopter to wings
	- updated user and head, head follows avatar over the worlds
	- renamed bone/ vnc/
	- moved wo/{body,face} to model/
	- updated url : saveToCacheAndOpen
	- updated hairs
	(2011/11/08, Philippe Dax)

-- 7.2.3
	- added prefs sex, head, skin, bust
	(2011/09/20, Philippe Dax)

-- 7.2.2
	- added option -t, --trace
	- added global localuser
	(2011/09/14, Philippe Dax)

-- 7.2.1
	- added object fire
	- added models butterfly, libellule, angel to wings
	- added models spires, bands, roses to dress
	- added models tophat, bowler, chinese, cardinal, country, clown to hat
	- added female to guy
	- added mysql to cloth
	- added colors to guy
	(2011/08/14, Philippe Dax)

-- 7.2.0
	- new layout
	- added bin and man directories
	- updated 3rdpart to extras
	- moved com to vreng
	(2011/07/30, Philippe Dax)

== 7.1
-- 7.1.0
	- reduced dependencies (ubit, png, zlib)
	- added builtin ubit
	- added builtin png
	- added builtin zlib
	(2011/07/28, Philippe Dax)

== 7.0
-- 7.0.2
	- added src/ubit
	- added src/3rdpart/ftgl
	(2011/07/19, Philippe Dax)

-- 7.0.1
	- fixed some g++ warnings
	- moved config/ to support/
	(2011/07/18, Philippe Dax, comeback after 2 years !)

-- 7.0.0
	- added flying bevavior for user
	- added command line option -a,--avatar to select an avatar model (man, guy)
	- added new user when enter into a new world
	- renamed spark to thunder
	- added wings object
	- updated user : can manage avatar model preferences
	- updated guy, now can be the avatar model by default
	- added flare-shine rendering
	- added teapot shape
	- added dress object
	- added hat object
	- added halo object
	- added cloth object
	- added sun object
	- added river object
	- added terrain object
	- added bubble object
	- added snow object

src/servers/vacs/vacs.cpp
//---------------------------------------------------------------------------
// VREng (Virtual Reality Engine)
// 
// Copyright (C) 1997-2008 Philippe Dax
// Telecom-ParisTech (Ecole Nationale Superieure des Telecommunications)
// 
// VREng is a free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public Licence as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
// 
// VREng is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//---------------------------------------------------------------------------
#include "sysdep.hpp"
#include "defaults.hpp"

#define VRENG_MPORT_INCR	10
#define WORLD_LEN		32
#define UNIVERSE_LEN		32
#define VACS_FILE		"worlds"
#define DEF_VACS_LOG		"/var/log/vacs.log"

// internal structure
struct _vacs {
  char url[URL_LEN];
  char world[WORLD_LEN];
  char channel[CHAN_LEN];
  char universe[UNIVERSE_LEN];
  uint32_t maddr;
  uint16_t mport;
  uint8_t ttl;
  uint8_t version;
  bool qualified;
  struct _vacs *next;
};


// global
bool debugvacs = 0;
char *log_file = NULL;
FILE *flog = NULL;
uint32_t next_maddr;
uint16_t next_mport;
uint8_t vre_version = VRE_VERSION;
struct _vacs *vacsList = NULL;
char vacs_file[PATH_LEN];
uint32_t cachesize = 0;


static void quitSig(int sig)
{
  fprintf(flog, "Got signal %d, aborting Vacs.\n", sig); fflush(flog);
  exit(sig);
}


/* Init signals */
void initSignals()
{
#ifndef WIN32
  signal(SIGILL, quitSig);
  signal(SIGTRAP, quitSig);
  signal(SIGFPE, quitSig);
  signal(SIGBUS, quitSig);
  signal(SIGSEGV, quitSig);
#endif // !WIN32
}


/* Compute date */
const char *date()
{
  time_t clock;
  struct tm *tm;
  static char datestr[20];

  clock = time(0);
  tm = localtime(&clock);
  sprintf(datestr, "%02d/%02d/%02d-%02d:%02d:%02d", tm->tm_mday, tm->tm_mon + 1, tm->tm_year % 100, tm->tm_hour, tm->tm_min, tm->tm_sec);
  return datestr;
}


/* Get a line from cache */
int getVacLine(char *line, FILE *fpvacs)
{
  do {
    if (fgets(line, BUFSIZ, fpvacs) == NULL)
      return -1;
  } while (line[0] == '#');
  line[strlen(line) -1] = 0;
  return strlen(line);
}


/* Extract universe name */
void getUniverse(const char *url, char *name)
{
  if (! strncmp(url, "http://", 7)) {
    char *u = strdup(url);
    char *p = strchr(u, '/');
    p++;
    p = strchr(p, '/');
    p++;
    char *p2 = strchr(p, '/');
    *p2 = 0;
    strcpy(name, p);      
    free(u);
  }
  else if (! strncmp(url, "file:", 5)) {
    strcpy(name, "local");      
  }
  else {
    strcpy(name, "none");      
  }
}


void getWorld(const char *url, char *name)
{
  char *u = strdup(url);
  char *p = strrchr(u, '.');
  *p = 0;
  p = strrchr(u, '/');
  p++;
  strcpy(name, p);
  free(u);
}


uint32_t getAddr(const char *channel)
{
  char *c = strdup(channel);
  char *p = strtok(c, "/");
#if HAVE_INET_PTON
  uint32_t addr;
  inet_pton(AF_INET, p, &addr);
#else
  uint32_t addr = inet_addr(p);
#endif
  free(c);
  return addr;
}


uint16_t getPort(const char *channel)
{
  char *c = strdup(channel);
  char *p = strtok(c, "/");
  p = strtok(NULL, "/");
  uint16_t port = atoi(p);
  free(c);
  return port;
}


uint8_t getTtl(const char *channel)
{
  uint8_t ttl = DEF_VRE_TTL;

  char *c = strdup(channel);
  char *p = strtok(c, "/");
  p = strtok(NULL, "/");
  p = strtok(NULL, "/");
  if (p) {
    ttl = atoi(p);
    if (ttl == 0)
      ttl = DEF_VRE_TTL;
  }
  free(c);
  return ttl;
}


/** Saves all the cache */
void saveCache()
{
  FILE *fpvacs;

  if ((fpvacs = fopen(vacs_file, "w")) == NULL) {
    perror("fopen write");
    return;
  }
  for (struct _vacs *w = vacsList->next; w ; w = w->next)
    fprintf(fpvacs, "%s\t%s\n", w->url, w->channel);
  fclose(fpvacs);
}


void sizeCache(int sdvre)
{
  char sizevacs[8];
  memset(sizevacs, 0, sizeof(sizevacs));
  sprintf(sizevacs, "%d", cachesize);
  write(sdvre, sizevacs, strlen(sizevacs));
}


void releaseCache(const char *world)
{
}


/** Resolves a world in its universe, returns channel if found */
bool resolveCache(const char *url, char *chanstr)
{
  fprintf(flog, "resolveCache: url=%s\n", url); fflush(flog);
  char world[WORLD_LEN], universe[UNIVERSE_LEN];

  getWorld(url, world);
  getUniverse(url, universe);

  for (struct _vacs *w = vacsList; w ; w = w->next) {
    if ((! strcmp(world, w->world)) && (! strcmp(universe, w->universe))) {
      char maddrstr[CHAN_LEN];
#if HAVE_INET_NTOP
      char dst[46];
      strcpy(maddrstr, inet_ntop(AF_INET, &w->maddr, dst, sizeof(dst)));
#else
      strcpy(maddrstr, inet_ntoa((struct in_addr &)w->maddr));
#endif
      sprintf(chanstr, "%s/%d/%d", maddrstr, w->mport, DEF_VRE_TTL);
      fprintf(flog, "resolveCache: chanstr=%s\n", chanstr); fflush(flog);
      return true;
    }
  }
  return false;
}


/** Adds vacs structure, appends to cache, returns new channel */
void addCache(const char *url, char *channel)
{
  fprintf(flog, "addCache: url=%s\n", url); fflush(flog);
  struct _vacs *vacs = (struct _vacs *) malloc(sizeof(_vacs));

  strcpy(vacs->url, url);			// url
  getWorld(vacs->url, vacs->world);		// extract world
  getUniverse(vacs->url, vacs->universe);	// extract universe

  if (! strncmp(vacs->url, "http://", 7)) {
    vacs->qualified = true;
  }
  else {
    vacs->qualified = false;
  }

  // maddr mport
  vacs->maddr = next_maddr;
  vacs->mport = next_mport;

  // channel
  char maddrstr[CHAN_LEN];
  memset(maddrstr, 0, sizeof(maddrstr));
#if HAVE_INET_NTOP
  char dst[46];
  strcpy(maddrstr, inet_ntop(AF_INET, &vacs->maddr, dst, sizeof(dst)));
#else
  strcpy(maddrstr, inet_ntoa((struct in_addr &)vacs->maddr));
#endif
  memset(vacs->channel, 0, sizeof(vacs->channel));
  sprintf(vacs->channel, "%s/%d/%d", maddrstr, vacs->mport, DEF_VRE_TTL); //vreng
  strcpy(channel, vacs->channel);
  sprintf(vacs->channel, "%s/%d", maddrstr, vacs->mport);	// record

  vacs->next = NULL;

  // add this entry at the end of the vacsList
  struct _vacs *wlast = NULL;
  for (struct _vacs *w = vacsList; w ; w = w->next) {
    wlast = w;
  }
  if (wlast) {
    wlast->next = vacs;
  }
  else
    fprintf(flog, "addCache: wlast NULL\n"); fflush(flog);

  // append this record to the file
  FILE *fpvacs;

  if (vacs->qualified) {
    if ((fpvacs = fopen(vacs_file, "a")) == NULL) {
      perror("fopen append");
      return;
    }
    fprintf(fpvacs, "%s\t%s\n", vacs->url, vacs->channel);
    fclose(fpvacs);
    fprintf(flog, "appended: %s\t%s\n", vacs->url, vacs->channel); fflush(flog);
  }
  else {
    fprintf(flog, "not appended: %s universe=%s\n", vacs->url, vacs->universe); fflush(flog);
  }

  // next values
  next_maddr = vacs->maddr + htonl(1);
  next_mport = vacs->mport + VRENG_MPORT_INCR;
}


/** Loads the cache file */
int loadCache()
{
  FILE *fpvacs;
  char line[BUFSIZ];
  char vacs_path[PATH_LEN];

  sprintf(vacs_path, "v%d/%s", VRE_VERSION, VACS_FILE);
  if ((fpvacs = fopen(vacs_path, "r")) == NULL) {
    perror("fopen read");
    exit(1);
  }

  // first record
  struct _vacs *vacs = (struct _vacs *) malloc(sizeof(_vacs));
  vacsList = vacs;
  strcpy(vacs->world, "Manager");
  *vacs->url = 0;
  sprintf(vacs->channel, "%s/%d/%d", VRENG_MADDR_BASE, VRENG_MPORT_BASE, DEF_VRE_TTL);
#if HAVE_INET_PTON
  inet_pton(AF_INET, VRENG_MADDR_BASE, &vacs->maddr);
#else
  vacs->maddr = inet_addr(VRENG_MADDR_BASE);
#endif
  vacs->mport = VRENG_MPORT_BASE;
  vacs->ttl = DEF_VRE_TTL;
  strcpy(vacs->universe, DEF_HTTP_SERVER);
  vacs->next = NULL;
  struct _vacs *prev = vacs;

  next_maddr = vacs->maddr + htonl(1);
  next_mport = vacs->mport + VRENG_MPORT_INCR;

  // other records
  int r = 0;
  while ((r = getVacLine(line, fpvacs) != -1)) {
    struct _vacs *vacs = (struct _vacs *) malloc(sizeof(_vacs));
    cachesize += strlen(line) + 1;

    // url
    char *p = strtok(line , " \t");
    if (!p)
      break;	// end of data

    strcpy(vacs->url, p);
    // world
    getWorld(vacs->url, vacs->world);
    // universe
    getUniverse(vacs->url, vacs->universe);

    // channel
    p = strtok(NULL, " \t");
    if (!p)
      break;	// end of data

    strcpy(vacs->channel, p);
    // maddr
    vacs->maddr = getAddr(p);
    // mport
    vacs->mport = getPort(p);
    vacs->ttl = getTtl(p);
    if (vacs->ttl == 0)
      vacs->ttl = DEF_VRE_TTL;

    // next values
    next_maddr = vacs->maddr + htonl(1);
    next_mport = vacs->mport + VRENG_MPORT_INCR;

    prev->next = vacs;
    vacs->next = NULL;
    prev = vacs;
    //fprintf(flog, "w=%s u=%s c=%s a=%x p=%d t=%d n=%x/%d\n", vacs->world, vacs->universe, vacs->channel, vacs->maddr, vacs->mport, vacs->ttl, next_maddr, next_mport); fflush(flog);
  }
  fclose(fpvacs);
  return cachesize;
}


void sendCache(int sdvre)
{
  char linevacs[URL_LEN + CHAN_LEN + 1];
  fprintf(flog, "sendCache:\n"); fflush(flog);

  for (struct _vacs *w = vacsList->next; w ; w = w->next) {
    memset(linevacs, 0, sizeof(linevacs));
    sprintf(linevacs, "%s %s/%d ", w->url, w->channel, DEF_VRE_TTL);
    write(sdvre, linevacs, strlen(linevacs));
#if HAVE_USLEEP
    //usleep(20000);	// tempo 20 ms
#endif
    //fprintf(flog, "%s %s/%d [%x]\n", w->url, w->channel, DEF_VRE_TTL, w->maddr); fflush(flog);
  }
  close(sdvre);
}


/** Handles requests from vreng client */
void parseRequest(int sdvre, const struct sockaddr_in *svre)
{
  char request[128], channel[CHAN_LEN];
  int len;

  memset(request, 0, sizeof(request));
  if ((len = read(sdvre, request, sizeof(request))) < 0) {
    perror("vacs: recv");
    return;
  }
  request[len] = '\0';
  fprintf(flog, "request: %s\n", request); fflush(flog);

  if (! strncmp(request, "list", 4)) {		// list
    char *p = request + 4;	// get the data version number
    uint8_t vers = atoi(p);
    if (vers)
      vre_version = vers;
    sprintf(vacs_file, "v%d/%s", vre_version, VACS_FILE);
    sendCache(sdvre);
  }
  else if (! strncmp(request, "resolve", 7)) {	// resolve
    char *url = strtok(request, " ");
    url = strtok(NULL, " ");
    if (resolveCache(url, channel) == 0) {
      addCache(url, channel);
    }
    channel[strlen(channel)] = 0;
    write(sdvre, channel, strlen(channel) + 1);
  }
  else if (! strncmp(request, "release", 7)) {	// release
    char *world = strtok(request, " ");
    world = strtok(NULL, " ");
    releaseCache(world);
  }
  else if (! strncmp(request, "size", 4)) {	// size
    char *world = strtok(request, " ");
    world = strtok(NULL, " ");
    sizeCache(sdvre);
  }
  else {					// unknown
    if (len != 0) {
      fprintf(flog, "unknown request: %s len=%d\n", request, len); fflush(flog);
    }
  }
  return;
}


/** Runs in the background */
void background()
{
  if (fork())
    exit(0);
}


/** Inits the server */
void initVacs()
{
  if (! debugvacs)
    background();

  initSignals();

  fprintf(flog, "%s started\n", date());
  fflush(flog);

  loadCache();

  int sdvacs, sdvre;
  struct sockaddr_in sa, svre;
  socklen_t slen = sizeof(struct sockaddr_in);

  if ((sdvacs = socket(PF_INET, SOCK_STREAM, 0)) < 0) {
    perror("socket");
    exit(1);
  }
  memset(&sa, 0, sizeof(sa));
  sa.sin_family = AF_INET;
  sa.sin_port = htons(VACS_PORT);
  sa.sin_addr.s_addr = htonl(INADDR_ANY);
  int one = 1;
  setsockopt(sdvacs, SOL_SOCKET, SO_REUSEADDR, (char *) &one, sizeof(one));
  if (bind(sdvacs, (struct sockaddr *) &sa, sizeof(sa)) < 0) {
    perror("binding");
    exit(1);
  }

  listen(sdvacs, 5);

  /* Server Main Loop */
  while (1) {
    memset(&svre, 0, sizeof(svre));
    if ((sdvre = accept(sdvacs, (struct sockaddr *) &svre, &slen)) < 0) {
      sleep(1);	// wait hoping that will be better next time...
      continue;
    } 
#if HAVE_INET_NTOP
    char dst[46];
    const char *pdst = inet_ntop(AF_INET, &svre.sin_addr, dst, sizeof(dst));
#else
    const char *pdst = inet_ntoa(svre.sin_addr);
#endif
    struct hostent *hp;
    if ((hp = gethostbyaddr((const char *) &svre.sin_addr,
                            sizeof(struct in_addr), AF_INET)) == NULL)
      perror("gethostbyaddr");
    fprintf(flog, "%s %s (%s) connected\n",
                  date(), (hp) ? hp->h_name : "<unresolved>", pdst);
    fflush(flog);

    parseRequest(sdvre, &svre);
    close(sdvre);
  } 
}


void usage(char *argv[])
{
  printf("usage: %s [-?] [-h] [-D] [-v] [-d vacsdir] [-l logfile]\n", argv[0]);
  exit(0);
} 

static char *vacs_dir = NULL;


/* Parse command line args */
void parseArgs(int argc, char *argv[])
{
  int c;

  while ((c = getopt(argc, argv, "?Dhvd:l:")) != -1) {
    switch (c) {
      case '?':
      case 'h': usage(argv); break;
      case 'D': debugvacs = true; break;
      case 'd': vacs_dir = optarg; break;
      case 'l': log_file = optarg; break;
      case 'v': printf("%s\n", PACKAGE_VERSION); exit(0);
    }
  }
  if (optind < argc) {
    usage(argv);
  }
}


int main(int argc, char *argv[])
{
  parseArgs(argc, argv);

  if (log_file) {
    if ((flog = fopen(log_file, "a")) == NULL) {
      fprintf(stderr, "can't open log_file %s\n", log_file);
    }
  }
  else
    flog = stderr;

  if (vacs_dir) {
    if (chdir(vacs_dir) < 0) {
      fprintf(stderr, "vacs: %s is not a directory\n", vacs_dir);
      exit(1);
    }
  }

  initVacs();
  return 0;
} 
